samples/extensions/sparse_image/sparse_image.h

```diff
@@ -0,0 +1,375 @@
+	{
+		if (left.expired())
+		{
+			return false;
+		}
+		else if (right.expired())
+		{
+			return true;
+		}
+		return left.lock()->available_offsets.size() > right.lock()->available_offsets.size();
+	};
+
+	struct VirtualTexture
+	{
+		VkImage      texture_image;
+		VkImageView  texture_image_view;
+		MemAllocInfo memory_allocations;
+
+		// Dimensions
+		size_t width;
+		size_t height;
+
+		// Number of bytes per page
+		size_t page_size;
+
+		uint8_t                    base_mip_level;
+		uint8_t                    mip_levels;
+		std::vector<MipProperties> mip_properties;
+
+		std::vector<std::vector<MipBlock>> current_mip_table;
+		std::vector<std::vector<MipBlock>> new_mip_table;
+
+		// Image containing a single, most detailed mip, allocated in the CPU memory, coppied to VRAM via staging buffer in update_and_generate()
+		std::unique_ptr<vkb::sg::Image> raw_data_image;
+
+		// Key table that includes data on which page is allocated to what memory block from the textureMemory vector
+		std::vector<PageTable> page_table;
+
+		// Set containing BLOCKS for which the required mip level has changed or/and its on-screen visibility changed
+		std::set<TextureBlock> texture_block_update_set;
+
+		// Set containing information which pages from the page_table should be updated (either loaded from CPU memory or blitted)
+		std::set<size_t> update_set;
+
+		// Vector of available memory pages (whole memory page pool is statically allocated at the beginning)
+		std::vector<size_t> available_memory_page_indices;
+
+		// Sparse-image-related format and memory properties
+		VkSparseImageFormatProperties   format_properties;
+		VkSparseImageMemoryRequirements memory_sparse_requirements;
+		VkMemoryRequirements            mem_requirements;
+
+		std::vector<VkSparseImageMemoryBind> sparse_image_memory_bind;
+	};
+
+	struct CalculateMipLevelData
+	{
+		std::vector<std::vector<Point>>    mesh;
+		std::vector<std::vector<MipBlock>> mip_table;
+
+		uint32_t vertical_num_blocks;
+		uint32_t horizontal_num_blocks;
+
+		uint8_t mip_levels;
+
+		std::vector<float> ax_vertical;
+		std::vector<float> ax_horizontal;
+
+		glm::mat4 mvp_transform;
+
+		VkExtent2D texture_base_dim;
+		VkExtent2D screen_base_dim;
+
+		CalculateMipLevelData(const glm::mat4 &mvp_transform, const VkExtent2D &texture_base_dim, const VkExtent2D &screen_base_dim, uint32_t vertical_num_blocks, uint32_t horizontal_num_blocks, uint8_t mip_levels);
+		CalculateMipLevelData();
+
+		void calculate_mesh_coordinates();
+		void calculate_mip_levels();
+	};
+
+	// UI related
+	bool color_highlight         = true;
+	bool color_highlight_changed = false;
+	bool memory_defragmentation  = true;
+	bool frame_counter_feature   = true;
+
+	size_t blocks_to_update_per_cycle = 10U;
+
+	size_t num_vertical_blocks   = 50U;
+	size_t num_horizontal_blocks = 50U;
+
+	size_t num_vertical_blocks_upd   = 50U;
+	size_t num_horizontal_blocks_upd = 50U;
```

Comment:
Maybe, similar to `color_highlight`, use some bool `num_[vertical|horizontal]_blocks_changed` instead?

Reply:
I tried that approach but faced some problems related to changing `num_vertical_blocks` value during the run-time. Having `num_vertical_blocks` changed once at the beginning of a cycle (in `calculate_mips_table()`) is a save solution.